### PR TITLE
Add messages for AddContactSchedulesToQueue service

### PR DESCRIPTION
**Pull request by wxmerkt**

### CODE REVIEW

File: msg/MpcStatistics.msg [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# This message provides essential diagnostics information
+
+time stamp
+
+uint16 num_iterations
+float32 problem_update_time
+float32 solve_time
+float32 final_cost
+float32 regularization_rate
```
Comment (cmastalli) [inline msg/MpcStatistics.msg:9]:
```suggestions
uint16 iterations
float32 total_time
float32 solve_time
float32 cost
float32 regularization
```

It is not clear to me what it is the difference between problem_update_time and solve_time.
Also, I think it is confusing terms final and rate.
Final might suggest we are solving the OC problem (actually it is more like a current cost in this sense)
Instead, rate suggest that it is a factor use to decrease or increase the regularization.

File: msg/MpcStatistics.msg [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+float32 step_length
+float32 feasibility
```
Comment (cmastalli) [inline msg/MpcStatistics.msg:11]:
```suggestion
 float32 dynamic_feasibility
```

This is optional, but we might want to publish different kind of feasibilities.
For instance, Crocoddyl in future will produce eq/ineq constraint feasibilities.
I personally like to have make a distinction as it is more insightful.

So, if you agree with me, we could also include these:
```c++
float32 equality_feasibility
float32 inequality_feasibility
```

**Review by cmastalli — COMMENTED** [review]
Conceptually everything looks alright for me.
I would just changes the names as suggested in my comment.

File: msg/MpcStatistics.msg (hunk 2 above)
Comment (wxmerkt) [reply msg/MpcStatistics.msg:11]: Thanks, that's a great point - I'll add them as _infeasibility_ as per usual nomenclature. We may want to have "primal_infeasibility" and "dual_infeasibility" for other types of algorithms, but that is not relevant for now.

File: msg/MpcStatistics.msg (hunk 1 above)
Comment (wxmerkt) [reply msg/MpcStatistics.msg]: I initially meant problem update time is the time taken to update all references, constraints, etc. - everything that is done _before_ calling solve. I've now changed it to total time though.

**Review by cmastalli — COMMENTED** [review]
@wxmerkt  -- the `ContactSchedule` message is tailored for quadruped robots.
It would be nice to modify it, but it requires substantial changes (especially in the caracal side)
Let me give you an overview of how I think we should do it

1. Create a `ContactManager` msg that contains a set of `ContactSchedule` msgs, each contact schedule per each leg.
2. Redefine the `ContactSchedule` msg in such a way it contains a set of `ContactPhase` msgs, each one defines timings and /or foothold target. The first phase should be active, the second inactive, and so on. Inactive phases need to define foothold target.
3. Create the `ContactPhase` msg as defined in 2.

This format mimics exactly the definition of a contact diagram introduced in Caracal.
It works for any kind of robot. 
Below, I provide the code for each message. 

- `ContactManager` msg
```c++
std_mgs/Header
whole_body_mpc_msgs/ContactSchedule[] contacts
```

- `ContactSchedule` msg
```c++
string name
whole_body_mpc_msgs/ContactPhase[] phase
```
- `ContactPhase` msg
```c++
float64 duration
float64 step_height
geometry_msg/Pose target
```

Finally, you'll need to modify your service as
```
whole_body_mpc_msgs/ContactManager
---
bool success
string message
```

Do you think you can handle this today or reasonably soon?

*Note that Caracal doesn't need the starting foothold, it uses the current one*
But, we could publish the starting footstep placement if you think it is relevant for display purpose.

**Comment by wxmerkt** [comment]
I've been looking through this, but I do not think that the suggested changes expose the same functionality -- it would require copying a gait generator outside in order to use the suggested interface (which I think makes sense for other use cases we may have too, but not the one currently implemented in caracal_ros). To achieve the other use case, we'd have to write additional code as well (for which I don't have the bandwidth right now until a use-case emerges).

The current proposition allows to directly add contact schedules using the gait generators mimicking their API. It also allows to modulate the gaits as with the gait config / hard-coded parameters for rapid experimentation. It's true that some of the gait types do not exist for systems outside of quadrupeds as noted in the message comments (and thus t_us, t_uds are meaningless), but otherwise the API can be equally be used with the biped generator (receiving node to be written). It wouldn't work for arm/manipulation contacts as-is, though.

Hence, after spending some time checking if we could adapt to it, it's not practical. What may make sense is to 
1. Rename the message as-is in this PR to `GaitGeneratorSchedule` / `AddGaitSchedulesToQueue` to unblock the merge - this way it's clearly limited to gaits / the gait generator and not a universal way of defining contact phases/diagrams
2. Add the proposed changes as additional messages (with naming `Contact...`). I won't have time to write the receiving/parsing part for those messages, so for the time being they would be defined messages without a use. 

What do you think?

**Comment by cmastalli** [comment]
I understand that this brings so many changes.

Your proposition seems the most viable solution given our constraints. We could proceed with it. Let me know when you have pushed it.

Furthermore, I would also ask you to create an issue regarding what it has been discussed here. This is for us to track future developments.

Btw, I do believe this proposition matches the requirements. We could definitely discuss it in this issue

**Comment by cmastalli** [comment]
Thank @wxmerkt!
I will review the other PRs as suggested

**Pull request merged**